File: 2024 October/Sort a k sorted doubly linked list/main.cpp
// Link: https://www.geeksforgeeks.org/problems/sort-a-k-sorted-doubly-linked-list/1

#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

class DLLNode
{
public:
    int data;
    DLLNode *next;
    DLLNode *prev;

    DLLNode(int val)
    {
        data = val;
        this->next = NULL;
        this->prev = NULL;
    }
};

void push(DLLNode **headRef, DLLNode **tailRef, int new_data)
{
    DLLNode *newNode = new DLLNode(new_data);

    if ((*tailRef) == NULL)
    {
        // If the list is empty, both head and tail point to the new node
        (*headRef) = newNode;
        (*tailRef) = newNode;
    }
    else
    {
        // Insert the new node at the end and update the tail
        newNode->prev = (*tailRef);
        (*tailRef)->next = newNode;
        (*tailRef) = newNode;
    }
}

void printList(DLLNode *head)
{
    if (head == NULL)
        return;

    while (head != NULL)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
}

DLLNode *createDoublyLinkedList(vector<int> &arr)
{
    DLLNode *head = NULL;
    DLLNode *tail = NULL;
    for (int i = 0; i < arr.size(); i++)
    {
        push(&head, &tail, arr[i]);
    }
    return head;
}

class Solution
{
public:
    DLLNode *sortAKSortedDLL(DLLNode *head, int k)
    {
        if (!head)
            return head;

        priority_queue<int, vector<int>, greater<int>> minHeap;
        DLLNode *newHead = NULL, *sortedTail = NULL, *current = head;

        for (int i = 0; i <= k && current; ++i)
        {
            minHeap.push(current->data);
            current = current->next;
        }

        while (!minHeap.empty())
        {
            int minValue = minHeap.top();
            minHeap.pop();

            if (!newHead)
            {
                newHead = new DLLNode(minValue);
                sortedTail = newHead;
            }
            else
            {
                sortedTail->next = new DLLNode(minValue);
                sortedTail->next->prev = sortedTail;
                sortedTail = sortedTail->next;
            }

            if (current)
            {
                minHeap.push(current->data);
                current = current->next;
            }
        }

        return newHead;
    }
};

int main()
{
    Solution obj;

    vector<int> arr = {3, 2, 1, 5, 6, 4};
    DLLNode *head = createDoublyLinkedList(arr);
    int k = 2;
    DLLNode *sortedHead = obj.sortAKSortedDLL(head, k);
    printList(sortedHead);

    arr = {5, 6, 7, 3, 4, 4};
    head = createDoublyLinkedList(arr);
    k = 3;
    sortedHead = obj.sortAKSortedDLL(head, k);
    printList(sortedHead);

    return 0;
}

File: 2024 October/Sort a k sorted doubly linked list/main.py
# Link: https://www.geeksforgeeks.org/problems/sort-a-k-sorted-doubly-linked-list/1

import heapq


# A node of the doubly linked list
class DLLNode:

    def __init__(self, val):
        self.data = val
        self.next = None
        self.prev = None


# Function to insert a node at the end of the doubly linked list
def push(tail, new_data):
    new_node = DLLNode(new_data)
    new_node.next = None
    new_node.prev = tail

    if tail is not None:
        tail.next = new_node

    return new_node


# Function to print nodes in a given doubly linked list
def printList(head):
    while head is not None:
        print(head.data, end=" ")
        head = head.next
    print()


def createDoublyLinkedList(arr):
    head = None
    tail = [None]
    for value in arr:
        push([head], tail, value)
        head = head if head is not None else tail[0]
    return head


class Solution:
    def sortAKSortedDLL(self, head, k):
        if not head:
            return head

        min_heap = []
        new_head = None
        sorted_tail = None
        current = head

        for i in range(k + 1):
            if current:
                heapq.heappush(min_heap, current.data)
                current = current.next

        while min_heap:
            min_value = heapq.heappop(min_heap)

            if new_head is None:
                new_head = DLLNode(min_value)
                sorted_tail = new_head
            else:
                sorted_tail.next = DLLNode(min_value)
                sorted_tail.next.prev = sorted_tail
                sorted_tail = sorted_tail.next

            if current:
                heapq.heappush(min_heap, current.data)
                current = current.next

        return new_head


if __name__ == "__main__":
    obj = Solution()

    arr = [3, 2, 1, 5, 6, 4]
    head = createDoublyLinkedList(arr)
    k = 2
    sorted_head = obj.sortAKSortedDLL(head, k)
    printList(sorted_head)

    arr = [5, 6, 7, 3, 4, 4]
    head = createDoublyLinkedList(arr)
    k = 3
    sorted_head = obj.sortAKSortedDLL(head, k)
    printList(sorted_head)
